src/components/cl/hier/bcast/bcast_2step.c

```diff
@@ -0,0 +1,255 @@
+            return i;
+        }
+    }
+    return UCC_RANK_INVALID;
+}
+
+static inline ucc_rank_t
+find_root_node_rank(ucc_rank_t root, ucc_cl_hier_team_t *cl_team)
+{
+    ucc_sbgp_t *sbgp       = cl_team->sbgps[UCC_HIER_SBGP_NODE].sbgp;
+    ucc_rank_t  i;
+
+    for (i = 0; i < sbgp->group_size; i++) {
+        if (ucc_ep_map_eval(sbgp->map, i) == root) {
+            return i;
+        }
+    }
+    return UCC_RANK_INVALID;
+}
+
+static ucc_status_t
+ucc_cl_hier_bcast_2step_init_schedule(ucc_base_coll_args_t *coll_args,
+                                      ucc_base_team_t      *team,
+                                      ucc_schedule_t **sched_p, int n_frags)
+{
+    ucc_cl_hier_team_t *cl_team   = ucc_derived_of(team, ucc_cl_hier_team_t);
+    ucc_team_t         *core_team = team->params.team;
+    ucc_coll_task_t    *tasks[2]  = {NULL, NULL};
+    ucc_rank_t root               = coll_args->args.root;
+    ucc_rank_t rank               = UCC_TL_TEAM_RANK(cl_team);
+    int root_on_local_node        = ucc_team_ranks_on_same_node(root, rank,
+                                                                core_team);
+    ucc_base_coll_args_t args       = *coll_args;
+    int                  n_tasks    = 0;
+    ucc_schedule_t      *schedule;
+    ucc_status_t         status;
+    int                  i;
+
+    schedule = &ucc_cl_hier_get_schedule(cl_team)->super.super;
+    if (ucc_unlikely(!schedule)) {
+        return UCC_ERR_NO_MEMORY;
+    }
+    status = ucc_schedule_init(schedule, &args, team);
+    if (ucc_unlikely(UCC_OK != status)) {
+        goto out;
+    }
+
+    if (n_frags > 1) {
+        args.max_frag_count =
+            ucc_buffer_block_count(args.args.src.info.count, n_frags, 0);
+        args.mask |= UCC_BASE_CARGS_MAX_FRAG_COUNT;
+    }
+
+    ucc_assert(SBGP_ENABLED(cl_team, NODE_LEADERS) ||
+               SBGP_ENABLED(cl_team, NODE));
+    if (SBGP_ENABLED(cl_team, NODE_LEADERS)) {
+        args.args.root = find_root_net_rank(
+            ucc_team_rank_host_id(root, core_team), cl_team);
+        status = ucc_coll_init(SCORE_MAP(cl_team, NODE_LEADERS), &args,
+                               &tasks[n_tasks]);
+        if (ucc_unlikely(UCC_OK != status)) {
+            goto out;
+        }
+        n_tasks++;
+    }
+
+    if (SBGP_ENABLED(cl_team, NODE)) {
+        args.args.root = root_on_local_node
+            ? find_root_node_rank(root, cl_team)
+            : core_team->topo->node_leader_rank_id;
+        status =
+            ucc_coll_init(SCORE_MAP(cl_team, NODE), &args, &tasks[n_tasks]);
+        if (ucc_unlikely(UCC_OK != status)) {
+            goto out;
+        }
+        n_tasks++;
+    }
+
+    ucc_task_subscribe_dep(&schedule->super, tasks[0],
+                           UCC_EVENT_SCHEDULE_STARTED);
+    ucc_schedule_add_task(schedule, tasks[0]);
+    if (n_tasks > 1) {
+        if (root == rank) {
+            ucc_task_subscribe_dep(&schedule->super, tasks[1],
```

Comment:
it's not correct, you can start 2 tasks (node leaders bcast and node bcast) in parallel only if root rank belongs to node leaders group.

Reply:
> it's not correct, you can start 2 tasks (node leaders bcast and node bcast) in parallel only if root rank belongs to node leaders group.

If root is not node leader then it has only 1 task, and will not get there

Reply:
ok, but node leader task should depend on node task in that case

Reply:
right.. this is fun. that is exactly why i had "first_task" before. and the condition above was correct actually to handle that case. i'll revert that last one change